pkg/volume/flexvolume/flexvolume.go

```diff
@@ -0,0 +1,324 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flexvolume
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+// This is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&flexVolumePlugin{nil}}
+}
+
+// FlexVolumePlugin object.
+type flexVolumePlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &flexVolumePlugin{}
+
+const (
+	flexVolumePluginName = "kubernetes.io/flexvolume"
+)
+
+// Init intializes the plugin.
+func (plugin *flexVolumePlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+// Name gets the name of the plugin.
+func (plugin *flexVolumePlugin) Name() string {
+	return flexVolumePluginName
+}
+
+// CanSupport checks whether the plugin can support the input volume spec.
+func (plugin *flexVolumePlugin) CanSupport(spec *volume.Spec) bool {
+	return (spec.PersistentVolume != nil && spec.PersistentVolume.Spec.FlexVolume != nil) ||
+		(spec.Volume != nil && spec.Volume.FlexVolume != nil)
+}
+
+// GetAccessModes gets the allowed access modes for this plugin.
+func (plugin *flexVolumePlugin) GetAccessModes() []api.PersistentVolumeAccessMode {
+	return []api.PersistentVolumeAccessMode{
+		api.ReadWriteOnce,
+	}
+}
+
+// NewBuilder is the builder routine to build the volume.
+func (plugin *flexVolumePlugin) NewBuilder(spec *volume.Spec, pod *api.Pod, _ volume.VolumeOptions, mounter mount.Interface) (volume.Builder, error) {
+	return plugin.newBuilderInternal(spec, pod, &flexVolumeUtil{}, mounter, exec.New())
```

Comment:
Are you planning to support SELinux with this volume type?  Currently the SELinux context of the volumes directory is passed to plugins via the `VolumeOptions` arg to `NewBuilder`.

You might also be interested in: #12944